Backpack/Skeleton/Classes/BPKSkeleton.swift

```diff
@@ -0,0 +1,175 @@
+        BPKColor.dynamicColor(withLightVariant: BPKColor.white, darkVariant: BPKColor.black)
+    private var customSize: CGSize = .zero
+    
+    public var type: BPKSkeletonType = .image {
+        didSet {
+            updateType()
+        }
+    }
+    
+    public var size: BPKSkeletonSize = .none
+    
+    public var style: BPKSkeletonStyle = .default {
+        didSet {
+            updateStyle()
+        }
+    }
+    
+    private var skeletonView: UIView = {
+        let view = UIView()
+        view.translatesAutoresizingMaskIntoConstraints = false
+        view.clipsToBounds = true
+        return view
+    }()
+    
+    public convenience init(
+        type: BPKSkeletonType,
+        size: BPKSkeletonSize = .default,
+        style: BPKSkeletonStyle = .default
+    ) {
+        self.init()
+        self.type = type
+        self.size = size
+        self.style = style
+        updateType()
+        setup()
+    }
+    
+    public convenience init(
+        type: BPKSkeletonType,
+        size: CGSize,
+        style: BPKSkeletonStyle = .default
+    ) {
+        self.init()
+        self.type = type
+        self.customSize = size
+        self.style = style
+        updateType()
+        setup()
+    }
+    
+    public override init(frame: CGRect) {
+        super.init(frame: frame)
+    }
+    
+    public required init?(coder: NSCoder) {
+        super.init(coder: coder)
+    }
+    
+    private func updateType() {
+        switch self.type {
+        case .bodytext:
+            self.skeletonView = BPKTextSkeleton(size: size == .none ? customSize : viewSize)
+        default:
+            self.skeletonView = BPKCommonSkeleton()
+        }
+    }
+    
+    internal func setup() {
+        addSubview(skeletonView)
+        let skeletonSize = size == .none ? customSize : viewSize
+        NSLayoutConstraint.activate([
+            skeletonView.leadingAnchor.constraint(equalTo: leadingAnchor),
+            skeletonView.topAnchor.constraint(equalTo: topAnchor),
+            skeletonView.trailingAnchor.constraint(equalTo: trailingAnchor),
+            skeletonView.bottomAnchor.constraint(equalTo: bottomAnchor),
+            
+            skeletonView.widthAnchor.constraint(equalToConstant: skeletonSize.width),
+            skeletonView.heightAnchor.constraint(equalToConstant: skeletonSize.height)
+        ])
+        
+        updateStyle()
+    }
+    
+    private func updateStyle() {
+        if type == .image {
+            skeletonView.layer.cornerRadius = cornerStyle
+        }
+        if type == .headline {
+            skeletonView.layer.cornerRadius = BPKCornerRadiusXs
+        }
+        if type == .circle && size != .none {
+            skeletonView.layer.cornerRadius = viewSize.height / 2.0
+        }
+        if type == .circle && size == .none {
+            skeletonView.layer.cornerRadius = min(customSize.height, customSize.width) / 2.0
+        }
+    }
+
+    public static func startShimmer(view: UIView) {
```

Comment:
Just fully realised this.
Can we make the skeleton view shimmer automatically? The use of a static method for this is not optimal also, since it can be called several times adding layers on top of other layers repeatedly.

I think this layer and animations could be part of the BPKSkeleton itself and then have an instance method (non-static) to set the skelenton's frame if needed

```swift
class BPKSkeleton {
  func setup() {
     startShimmer()
  }
  func startShimmer() {
    self.gradientLayer = makeLayer() // all the logic here
    self.layer.addSublayer(gradientLayer) // Called on self, aka the BPKSkeleton
  }

  public func updateSize(frame) {
    self.gradientLayer.frame = frame
  }
}
```

Reply:
using layoutsubviews to update the layer's frame is recommended, because it will automatically honour the skeleton's size, set from autolayouts on the user's side.

Reply:
I see, I did this previously, but it not meet the requirement. I did this before and it looks like this: 

https://user-images.githubusercontent.com/13347837/186140640-bd9ade76-883b-4e9f-9b5a-309a2c034b4d.mp4


Because this way will add shimmer animation to every Skeleton instance, image skeleton has a shimmer animation, circle has a shimmer animation, etc.

But UI requirement is only 1 animation in the parent view. Here is the [Figma ](https://www.figma.com/proto/SeFyp0yJXEFUVAlAmrU7xc/Skeleton-Loader?page-id=2%3A282&node-id=728%3A6271&viewport=1462%2C31%2C1.8&scaling=contain&starting-point-node-id=4%3A5262)

Reply:
ah, indeed I didn't know that was the requirement!
My suggestion would not work as intended then